### PR TITLE
IGNITE-21445 New implementation for local IndexQuery

### PR DESCRIPTION
**Pull request by oleg-vlsk**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java [hunk 1]
```diff
@@ -625,7 +629,7 @@ private <R> CacheQueryFuture<R> execute0(@Nullable IgniteReducer<T, R> rmtReduce
         GridCloseableIterator it;
 
         if (loc)
-            it = qryMgr.scanQueryLocal(this, true);
+            it = qryMgr.queryLocal(this, true);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java:632]: It looks strange for non-local query we call `scanQueryDistributed` while for local it calls generic method. Let's separate `scanQueryLocal` and `indexQueryLocal`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 2]
```diff
@@ -1411,17 +1411,19 @@ else if (rmvIter)
...
     /**
-     * Process local scan query.
+     * Process local query with results returned in an iterator.
      *
      * @param qry Query.
      * @param updateStatistics Update statistics flag.
      */
     @SuppressWarnings({"unchecked"})
-    protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
+    protected GridCloseableIterator queryLocal(final GridCacheQueryAdapter qry,
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1420]: Let's create 2 methods - `scanQueryLocal` and `indexQueryLocal`. They should reuse some common code.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java [hunk 3]
```diff
@@ -607,7 +611,7 @@ private <R> CacheQueryFuture<R> execute0(@Nullable IgniteReducer<T, R> rmtReduce
                 }
             }
 
-            return new GridEmptyCloseableIterator();
+            throw new IgniteException(new ClusterGroupEmptyException());
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java:614]: If this kind of Exception is not thrown here then the IndexQueryLocalTest#testClientNodeReplicatedCache() will fail.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java [hunk 4]
```diff
@@ -579,21 +580,24 @@ private <R> CacheQueryFuture<R> execute0(@Nullable IgniteReducer<T, R> rmtReduce
...
+    @Override public GridCloseableIterator executeQueryWithIterator() throws IgniteCheckedException {
+        assert type == SCAN || type == INDEX : "Wrong processing of query: " + type;
 
         GridDhtCacheAdapter<?, ?> cacheAdapter = cctx.isNear() ? cctx.near().dht() : cctx.dht();
 
         Set<Integer> lostParts = cacheAdapter.topology().lostPartitions();
 
         if (!lostParts.isEmpty()) {
             if (part == null || lostParts.contains(part)) {
-                throw new CacheException(new CacheInvalidStateException("Failed to execute query because cache partition " +
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java:591]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java [hunk 5]
```diff
@@ -818,8 +840,8 @@ private QueryCursor<Cache.Entry<K, V>> queryContinuous(AbstractContinuousQuery q
                 return (FieldsQueryCursor<R>)ctx.kernalContext().query().querySqlFields(ctx, (SqlFieldsQuery)qry,
                     null, keepBinary, true).get(0);
 
-            if (qry instanceof ScanQuery)
-                return query((ScanQuery)qry, null, projection(qry.isLocal()));
+            if (qry instanceof ScanQuery || (qry instanceof IndexQuery && qry.isLocal()))
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java [hunk 6]
```diff
@@ -475,23 +475,45 @@ private <T, R> QueryCursor<R> query(
...
+
+            qryType = GridCacheQueryType.SCAN;
+
+            IgniteBiPredicate<K, V> p = scanQry.getFilter();
+
+            cacheQry = ctx.queries().createScanQuery(
+                p, transformer, scanQry.getPartition(), isKeepBinary, qry.isLocal(), null);
+        }
+        else {
+            IndexQuery idxQryLoc = (IndexQuery)qry;
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java (hunk 1 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java (hunk 2 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryType.java [hunk 7]
```diff
@@ -43,5 +43,8 @@ public enum CacheQueryType {
     CONTINUOUS,
 
     /** SPI query. */
-    SPI
+    SPI,
+
+    /** Index query. */
+    INDEX
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryType.java:49]: Looks like now we need a test for new QueryType

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 8]
```diff
@@ -1483,6 +1487,62 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
+        return queryFunc(() -> {
+            try {
+                IgniteClosure transformer = qry.transform();
+
+                injectResources(transformer);
+
+                return scanIterator(qry, transformer, true);
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1509]: Got rid of 'try-catch' in scanQueryLocal and indexQueryLocal by using IgniteThrowableSupplier.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 9]
```diff
@@ -1434,22 +1439,25 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
                 cctx.gridEvents().record(new CacheQueryExecutedEvent<>(
                     locNode,
-                    "Scan query executed.",
+                    qry.type() + " query executed.",
                     EVT_CACHE_QUERY_EXECUTED,
-                    CacheQueryType.SCAN.name(),
+                    qry.type() == SCAN ? CacheQueryType.SCAN.name() : CacheQueryType.INDEX.name(),
                     namex,
-                    null,
+                    clsName,
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java]: Before this code was used for scan queries only, so the class name was irrelevant for cache query events and therefore was always ‘null’. Now this code processes both scan and index queries, and with index queries we need to record the name of the queried class to the event.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryType.java (hunk 7 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryType.java:49]: Test added, new functionality for recording index query events added.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 10]
```diff
@@ -1434,22 +1439,25 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
             new InternalScanFilter<>(qry.scanFilter()) : null;
 
         try {
-            assert qry.type() == SCAN;
+            assert qry.type() == SCAN || qry.type() == INDEX;
 
             if (log.isDebugEnabled())
-                log.debug("Running local SCAN query: " + qry);
-
-            final String taskName = cctx.kernalContext().task().resolveTaskName(qry.taskHash());
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1442]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 11]
```diff
@@ -1434,22 +1439,25 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
+
+                final String clsName = qry.queryClassName();
+
                 cctx.gridEvents().record(new CacheQueryExecutedEvent<>(
                     locNode,
-                    "Scan query executed.",
+                    qry.type() + " query executed.",
                     EVT_CACHE_QUERY_EXECUTED,
-                    CacheQueryType.SCAN.name(),
+                    qry.type() == SCAN ? CacheQueryType.SCAN.name() : CacheQueryType.INDEX.name(),
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 12]
```diff
@@ -1458,11 +1466,7 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
             }
 
-            IgniteClosure transformer = qry.transform();
-
-            injectResources(transformer);
-
-            GridCloseableIterator it = scanIterator(qry, transformer, true);
+            GridCloseableIterator it = qrySupplier.get();
 
             updateStatistics = false;
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1505]: Agreed to leave it as it is and revisit it later on.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java]: Got rid of 'try-catch' in scanQueryLocal and indexQueryLocal by using IgniteThrowableSupplier.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 13]
```diff
@@ -3427,4 +3503,56 @@ IgniteBiPredicate<K, V> scanFilter() {
             return scanFilter;
         }
     }
+
+    /** */
+    public static final class IndexQueryIterator<K, V> extends GridCloseableIteratorAdapter<Object> {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:3508]: private

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 14]
```diff
@@ -1499,6 +1502,79 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
+
+            if (!lostParts.isEmpty()) {
+                if (part == null || lostParts.contains(part)) {
+                    throw new CacheException(new CacheInvalidStateException("Failed to execute query because cache partition " +
+                        "has been lost [cacheName=" + cctx.name() +
+                        ", part=" + (part == null ? lostParts.iterator().next() : part) + ']'));
+                }
+            }
+
+            if (qry.nodes().isEmpty())
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1535]: IndexQueryLocalTest#testClientNodeReplicatedCache will fail if we don't throw a certain exception after this check.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 15]
```diff
@@ -1499,6 +1502,79 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
+
+            cctx.checkSecurity(SecurityPermission.CACHE_READ);
+
+            GridDhtCacheAdapter<?, ?> cacheAdapter = cctx.isNear() ? cctx.near().dht() : cctx.dht();
+
+            Set<Integer> lostParts = cacheAdapter.topology().lostPartitions();
+
+            Integer part = qry.partition();
+
+            if (!lostParts.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1527]: Lost partitions check removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 16]
```diff
@@ -1499,6 +1502,79 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
+                        "has been lost [cacheName=" + cctx.name() +
+                        ", part=" + (part == null ? lostParts.iterator().next() : part) + ']'));
+                }
+            }
+
+            if (qry.nodes().isEmpty())
+                throw new IgniteException(new ClusterGroupEmptyException());
+
+            if (cctx.deploymentEnabled())
+                cctx.deploy().registerClasses(qry.scanFilter());
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1539]: Code removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 17]
```diff
@@ -1499,6 +1502,79 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
+            if (part != null)
+                parts = new int[] {part};
+
+            IndexQueryResult<K, V> idxQryRes = qryProc.queryIndex(cacheName, qry.queryClassName(), qry.idxQryDesc(),
+                qry.scanFilter(), filter(qry, parts, parts != null), qry.keepBinary(), qry.taskHash());
+
+            return new IndexQueryIterator(idxQryRes.iter());
+        }
+        catch (Exception e) {
+            throw e;
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1571]: Catch section removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java (hunk 13 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:3508]: Changed to `private`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java [hunk 18]
```diff
@@ -557,6 +558,17 @@ else if (query instanceof IndexQuery) {
...
 
+            if (query.isLocal()) {
+                final GridCloseableIterator iter = ctx.kernalContext().query().executeQuery(GridCacheQueryType.INDEX,
+                    cacheName, ctx, new IgniteOutClosureX<GridCloseableIterator>() {
+                        @Override public GridCloseableIterator applyx() throws IgniteCheckedException {
+                            return ctx.queries().indexQueryLocal((GridCacheQueryAdapter)qry);
+                        }
+                    }, true);
+
+                return new QueryCursorImpl<>(iter);
```
[the remark this answers is not in the document]
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java:569]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 19]
```diff
@@ -1499,6 +1500,59 @@ protected GridCloseableIterator scanQueryLocal(final GridCacheQueryAdapter qry,
...
+    public GridCloseableIterator indexQueryLocal(final GridCacheQueryAdapter qry) throws IgniteCheckedException {
+        if (!enterBusy())
+            throw new IllegalStateException("Failed to process query request (grid is stopping).");
+
+        try {
+            assert qry.type() == INDEX : "Wrong processing of query: " + qry.type();
+
+            cctx.checkSecurity(SecurityPermission.CACHE_READ);
+
+            if (qry.nodes().isEmpty())
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1519]: If the query is local there is no need to check the nodes.

Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java]: The check is changed to `cctx.localNode().isClient()`.

**Review by timoninmaxim — APPROVED** [review]

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**